File: test/SmtpTest.java
import com.icegreen.greenmail.util.GreenMail;
import com.icegreen.greenmail.util.ServerSetup;
import jakarta.mail.internet.MimeMultipart;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import java.util.HashMap;
import java.util.Map;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Exercises the mail path against a real SMTP server rather than a mock: the transport
 * connects, authenticates, and the message that actually arrives is inspected.
 */
class SmtpTest {

  private static final Event EVENT = new Event(
      "20260917@jug-da.de",
      "Supercharging Java with Quarkus (Marco Klaassen)",
      "Supercharging Java with Quarkus",
      "abstract",
      "Marco Klaassen",
      "marco",
      "MaibornWolff GmbH, Darmstadt",
      "https://www.jug-da.de/2026/09/Quarkus-LLMs/",
      "2026-09-17T18:30:00",
      false,
      false);

  private GreenMail smtp;

  @BeforeEach
  void startServer() {
    var setup = new ServerSetup(0, "127.0.0.1", ServerSetup.PROTOCOL_SMTP);
    setup.setServerStartupTimeout(5000);
    smtp = new GreenMail(setup);
    smtp.setUser("orga@jug-da.de", "jugda", "s3cret");
    smtp.start();
  }

  @AfterEach
  void stopServer() {
    smtp.stop();
  }

  private Config.Mail config(String password) {
    var env = new HashMap<String, String>();
    env.put("EVENTS_URL", "https://example.invalid/events.json");
    env.put("MAIL_TO", "liste@jug-da.de");
    env.put("MAIL_FROM", "orga@jug-da.de");
    env.put("SMTP_HOST", "127.0.0.1");
    env.put("SMTP_PORT", String.valueOf(smtp.getSmtp().getPort()));
    env.put("SMTP_USER", "jugda");
    env.put("SMTP_PASS", password);
    return Config.load(env).mail();
  }

  @Test
  @DisplayName("delivers a real message through a plain authenticated SMTP server")
  void deliversRealMessage() throws Exception {
    var milestone = Mail.due(EVENT, 7).orElseThrow();
    var messageId = Mail.send(EVENT, milestone, config("s3cret"));
    assertNotNull(messageId);

    assertTrue(smtp.waitForIncomingEmail(5000, 1), "no message arrived");
    var received = smtp.getReceivedMessages();
    assertEquals(1, received.length);

    var message = received[0];
    assertEquals("orga@jug-da.de", message.getFrom()[0].toString());
    assertEquals("liste@jug-da.de", message.getAllRecipients()[0].toString());
    assertEquals("17.09.2026: Supercharging Java with Quarkus (Marco Klaassen)", message.getSubject());

    assertTrue(message.getContentType().toLowerCase().contains("multipart/alternative"), message.getContentType());
    var body = (MimeMultipart) message.getContent();
    assertEquals(2, body.getCount(), "expected a text and an html part");
    assertTrue(body.getBodyPart(0).getContentType().toLowerCase().contains("text/plain"));
    assertTrue(body.getBodyPart(1).getContentType().toLowerCase().contains("text/html"));

    var text = body.getBodyPart(0).getContent().toString();
    assertTrue(text.contains("Donnerstag, 17. September 2026"), text);
    assertTrue(text.contains("18:30"), text);
    // The link target has to survive into the plain-text part.
    assertTrue(text.contains("https://www.jug-da.de/2026/09/Quarkus-LLMs/"), text);
  }

  @Test
  @DisplayName("the announcement is a different template with its own subject prefix")
  void announcementDiffers() throws Exception {
    var milestone = Mail.due(EVENT, 28).orElseThrow();
    Mail.send(EVENT, milestone, config("s3cret"));

    assertTrue(smtp.waitForIncomingEmail(5000, 1), "no message arrived");
    var message = smtp.getReceivedMessages()[0];
    assertTrue(message.getSubject().startsWith("Ankündigung für 17.09.2026:"), message.getSubject());

    var body = (MimeMultipart) message.getContent();
    assertTrue(body.getBodyPart(0).getContent().toString().contains("Die Anmeldung ist jetzt ebenfalls geöffnet"));
  }

  @Test
  @DisplayName("rejects a wrong password rather than silently dropping the mail")
  void wrongPasswordFailsLoudly() {
    var milestone = Mail.due(EVENT, 7).orElseThrow();
    assertThrows(Exception.class, () -> Mail.send(EVENT, milestone, config("wrong")));
    assertEquals(0, smtp.getReceivedMessages().length);
  }
}

File: test/smtp.test.js
// Proves the mail path end to end against a real SMTP server: the transport connects,
// authenticates with whatever SMTP_USER/SMTP_PASS were configured, and the message that
// arrives carries the right envelope, subject and both body parts.

import assert from 'node:assert/strict';
import { after, before, test } from 'node:test';
import { SMTPServer } from 'smtp-server';
import * as config from '../src/config.js';
import * as mail from '../src/mail.js';

const received = [];
let server;
let port;

const listen = (srv) => new Promise((resolve) => srv.listen(0, '127.0.0.1', () => resolve(srv.server.address().port)));

before(async () => {
  server = new SMTPServer({
    authOptional: false,
    disabledCommands: ['STARTTLS'],
    onAuth(auth, _session, callback) {
      if (auth.username === 'jugda' && auth.password === 's3cret') return callback(null, { user: auth.username });
      return callback(new Error('bad credentials'));
    },
    onData(stream, session, callback) {
      let raw = '';
      stream.on('data', (chunk) => { raw += chunk; });
      stream.on('end', () => {
        received.push({ envelope: session.envelope, raw });
        callback();
      });
    },
  });
  port = await listen(server);
});

after(() => server.close());

const env = (overrides = {}) => ({
  EVENTS_URL: 'https://example.invalid/events.json',
  MAIL_TO: 'liste@jug-da.de',
  MAIL_FROM: 'orga@jug-da.de',
  SMTP_HOST: '127.0.0.1',
  SMTP_PORT: String(port),
  SMTP_USER: 'jugda',
  SMTP_PASS: 's3cret',
  ...overrides,
});

const event = {
  uid: '20260917@jug-da.de',
  start: '2026-09-17T18:30:00',
  summary: 'Supercharging Java with Quarkus (Marco Klaassen)',
  location: 'MaibornWolff GmbH, Darmstadt',
  url: 'https://www.jug-da.de/2026/09/Quarkus-LLMs/',
};

test('delivers a real message through a plain authenticated SMTP server', async () => {
  const cfg = config.load(env());
  const milestone = mail.due(event, 7);

  const messageId = await mail.send(event, milestone, cfg.mail);
  assert.ok(messageId, 'expected a message id back from the server');

  assert.equal(received.length, 1);
  const { envelope, raw } = received[0];
  assert.equal(envelope.mailFrom.address, 'orga@jug-da.de');
  assert.deepEqual(envelope.rcptTo.map((r) => r.address), ['liste@jug-da.de']);

  // Subject is MIME-encoded because of the umlaut, so assert on the decoded form.
  const subject = raw.match(/^Subject: (.*)$/m)[1];
  const decoded = Buffer.from(subject.match(/\?B\?(.+)\?=/)?.[1] ?? '', 'base64').toString('utf-8');
  assert.equal(decoded || subject, '17.09.2026: Supercharging Java with Quarkus (Marco Klaassen)');

  assert.match(raw, /multipart\/alternative/);
  assert.match(raw, /text\/plain/);
  assert.match(raw, /text\/html/);
});

test('rejects a wrong SMTP password rather than silently dropping the mail', async () => {
  // A fresh module registry, because mail.js memoises its transport.
  const fresh = await import(`../src/mail.js?wrong-pass`);
  const cfg = config.load(env({ SMTP_PASS: 'wrong' }));
  await assert.rejects(() => fresh.send(event, fresh.due(event, 7), cfg.mail), /Invalid login|bad credentials/i);
});
